File: frontend/lib/services/announcement.service.ts
import { apiClient } from '@/lib/api/client';

export interface Announcement {
  id: string;
  title: string;
  content: string;
  category: AnnouncementCategory;
  priority: AnnouncementPriority;
  status: AnnouncementStatus;
  targetAudience: TargetAudience;
  targetDepartmentIds?: string[];
  targetEmployeeIds?: string[];
  publishedAt: string;
  expiresAt?: string;
  isPinned: boolean;
  publishedBy?: string;
  publishedByName?: string;
  attachmentUrl?: string;
  readCount: number;
  isRead?: boolean;
  createdAt: string;
  updatedAt: string;
  wallPostId?: string; // Reference to wall post for reactions/comments
  hasReacted?: boolean; // Whether current user has reacted
}

export type AnnouncementCategory =
  | 'GENERAL'
  | 'POLICY_UPDATE'
  | 'EVENT'
  | 'HOLIDAY'
  | 'ACHIEVEMENT'
  | 'URGENT'
  | 'BENEFIT'
  | 'TRAINING'
  | 'SOCIAL'
  | 'IT_MAINTENANCE'
  | 'HEALTH_SAFETY'
  | 'OTHER';

export type AnnouncementPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export type AnnouncementStatus = 'DRAFT' | 'SCHEDULED' | 'PUBLISHED' | 'EXPIRED' | 'ARCHIVED';

export type TargetAudience =
  | 'ALL_EMPLOYEES'
  | 'SPECIFIC_DEPARTMENTS'
  | 'SPECIFIC_EMPLOYEES'
  | 'MANAGERS_ONLY'
  | 'NEW_JOINERS';

export interface CreateAnnouncementRequest {
  title: string;
  content: string;
  category: AnnouncementCategory;
  priority: AnnouncementPriority;
  targetAudience: TargetAudience;
  targetDepartmentIds?: string[];
  targetEmployeeIds?: string[];
  publishedAt?: string;
  expiresAt?: string;
  isPinned?: boolean;
  sendEmail?: boolean;
  attachmentUrl?: string;
}

export interface Page<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

class AnnouncementService {
  async getAllAnnouncements(page: number = 0, size: number = 10): Promise<Page<Announcement>> {
    const response = await apiClient.get<Page<Announcement>>('/announcements', {
      params: { page, size },
    });
    return response.data;
  }

  async getActiveAnnouncements(employeeId: string, page: number = 0, size: number = 10): Promise<Page<Announcement>> {
    const response = await apiClient.get<Page<Announcement>>('/announcements/active', {
      params: { employeeId, page, size },
    });
    return response.data;
  }

  async getPinnedAnnouncements(): Promise<Announcement[]> {
    const response = await apiClient.get<Announcement[]>('/announcements/pinned');
    return response.data;
  }

  async getAnnouncement(id: string, employeeId?: string): Promise<Announcement> {
    const response = await apiClient.get<Announcement>(`/announcements/${id}`, {
      params: employeeId ? { employeeId } : undefined,
    });
    return response.data;
  }

  async createAnnouncement(request: CreateAnnouncementRequest): Promise<Announcement> {
    const response = await apiClient.post<Announcement>('/announcements', request);
    return response.data;
  }

  async updateAnnouncement(id: string, request: CreateAnnouncementRequest): Promise<Announcement> {
    const response = await apiClient.put<Announcement>(`/announcements/${id}`, request);
    return response.data;
  }

  async deleteAnnouncement(id: string): Promise<void> {
    await apiClient.delete(`/announcements/${id}`);
  }

  async markAsRead(announcementId: string, employeeId: string): Promise<void> {
    await apiClient.post(`/announcements/${announcementId}/read`, null, {
      params: { employeeId },
    });
  }
}

export const announcementService = new AnnouncementService();

// Helper functions
export function getCategoryColor(category: AnnouncementCategory): string {
  const colors: Record<AnnouncementCategory, string> = {
    GENERAL: 'bg-accent-100 text-accent-800',
    POLICY_UPDATE: 'bg-accent-100 text-accent-800',
    EVENT: 'bg-success-100 text-success-800',
    HOLIDAY: 'bg-warning-100 text-warning-800',
    ACHIEVEMENT: 'bg-accent-100 text-accent-800',
    URGENT: 'bg-danger-100 text-danger-800',
    BENEFIT: 'bg-accent-100 text-accent-800',
    TRAINING: 'bg-accent-100 text-accent-800',
    SOCIAL: 'bg-warning-100 text-warning-800',
    IT_MAINTENANCE: 'bg-[var(--bg-surface)] text-[var(--text-primary)]',
    HEALTH_SAFETY: 'bg-success-100 text-success-800',
    OTHER: 'bg-[var(--bg-surface)] text-[var(--text-primary)]',
  };
  return colors[category] || colors.OTHER;
}

export function getPriorityColor(priority: AnnouncementPriority): string {
  const colors: Record<AnnouncementPriority, string> = {
    LOW: 'bg-[var(--bg-surface)] text-[var(--text-primary)]',
    MEDIUM: 'bg-accent-100 text-accent-800',
    HIGH: 'bg-warning-100 text-warning-800',
    CRITICAL: 'bg-danger-100 text-danger-800',
  };
  return colors[priority] || colors.MEDIUM;
}

export function getCategoryLabel(category: AnnouncementCategory): string {
  const labels: Record<AnnouncementCategory, string> = {
    GENERAL: 'General',
    POLICY_UPDATE: 'Policy Update',
    EVENT: 'Event',
    HOLIDAY: 'Holiday',
    ACHIEVEMENT: 'Achievement',
    URGENT: 'Urgent',
    BENEFIT: 'Benefits',
    TRAINING: 'Training',
    SOCIAL: 'Social',
    IT_MAINTENANCE: 'IT Maintenance',
    HEALTH_SAFETY: 'Health & Safety',
    OTHER: 'Other',
  };
  return labels[category] || category;
}

File: backend/src/main/java/com/hrms/application/announcement/service/AnnouncementService.java
package com.hrms.application.announcement.service;

import com.hrms.api.announcement.dto.AnnouncementDto;
import com.hrms.api.announcement.dto.CreateAnnouncementRequest;
import com.hrms.api.wall.dto.CreatePostRequest;
import com.hrms.api.wall.dto.WallPostResponse;
import com.hrms.application.common.service.ContentViewService;
import com.hrms.application.wall.service.WallService;
import com.hrms.common.exception.ResourceNotFoundException;
import com.hrms.common.security.SecurityContext;
import com.hrms.common.security.TenantContext;
import com.hrms.domain.announcement.Announcement;
import com.hrms.domain.announcement.Announcement.*;
import com.hrms.domain.announcement.AnnouncementRead;
import com.hrms.domain.common.ContentView.ContentType;
import com.hrms.domain.wall.model.WallPost;
import com.hrms.infrastructure.announcement.repository.AnnouncementRepository;
import com.hrms.infrastructure.announcement.repository.AnnouncementReadRepository;
import com.hrms.infrastructure.employee.repository.EmployeeRepository;
import com.hrms.infrastructure.wall.repository.PostReactionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class AnnouncementService {

    private final AnnouncementRepository announcementRepository;
    private final AnnouncementReadRepository announcementReadRepository;
    private final EmployeeRepository employeeRepository;
    private final ContentViewService contentViewService;
    private final WallService wallService;
    private final PostReactionRepository postReactionRepository;

    @Transactional
    public AnnouncementDto createAnnouncement(CreateAnnouncementRequest request) {
        UUID tenantId = TenantContext.getCurrentTenant();
        UUID userId = SecurityContext.getCurrentUserId();
        // BUG-FIX: Use employeeId for employee table lookups (userId != employeeId)
        UUID employeeId = SecurityContext.getCurrentEmployeeId();
        UUID lookupId = employeeId != null ? employeeId : userId;

        String publisherName = "System";
        if (lookupId != null) {
            publisherName = employeeRepository.findByIdAndTenantId(lookupId, tenantId)
                    .map(emp -> emp.getFirstName() + " " + emp.getLastName())
                    .orElse("System");
        }

        Announcement announcement = Announcement.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .category(request.getCategory() != null ? request.getCategory() : AnnouncementCategory.GENERAL)
                .priority(request.getPriority() != null ? request.getPriority() : AnnouncementPriority.MEDIUM)
                .targetAudience(request.getTargetAudience() != null ? request.getTargetAudience() : TargetAudience.ALL_EMPLOYEES)
                .targetDepartmentIds(request.getTargetDepartmentIds() != null ? request.getTargetDepartmentIds() : Set.of())
                .targetEmployeeIds(request.getTargetEmployeeIds() != null ? request.getTargetEmployeeIds() : Set.of())
                .expiresAt(request.getExpiresAt())
                .isPinned(request.getIsPinned() != null && request.getIsPinned())
                .sendEmail(request.getSendEmail() != null && request.getSendEmail())
                .attachmentUrl(request.getAttachmentUrl())
                .requiresAcceptance(request.getRequiresAcceptance() != null && request.getRequiresAcceptance())
                .build();

        announcement.setTenantId(tenantId);
        announcement.publish(userId, publisherName);

        announcement = announcementRepository.save(announcement);

        // Create wall post for social features (reactions/comments)
        // BUG-FIX: Use employeeId (not userId) since WallService looks up the employee table
        try {
            CreatePostRequest wallPostRequest = new CreatePostRequest();
            wallPostRequest.setType(WallPost.PostType.POST);
            wallPostRequest.setContent(announcement.getContent());
            wallPostRequest.setVisibility(WallPost.PostVisibility.ORGANIZATION);

            WallPostResponse wallPost = wallService.createPost(wallPostRequest, lookupId);
            announcement.setWallPostId(wallPost.getId());
            announcement = announcementRepository.save(announcement);
            log.info("Created wall post {} for announcement {}", wallPost.getId(), announcement.getId());
        } catch (Exception e) { // Intentional broad catch — notification delivery error boundary
            log.error("Failed to create wall post for announcement {}: {}", announcement.getId(), e.getMessage());
            // Don't fail the announcement if wall post creation fails
        }

        log.info("Created announcement: {} by {}", announcement.getTitle(), publisherName);

        return AnnouncementDto.fromEntity(announcement);
    }

    @Transactional
    public AnnouncementDto updateAnnouncement(UUID announcementId, CreateAnnouncementRequest request) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        announcement.setTitle(request.getTitle());
        announcement.setContent(request.getContent());
        if (request.getCategory() != null) announcement.setCategory(request.getCategory());
        if (request.getPriority() != null) announcement.setPriority(request.getPriority());
        if (request.getTargetAudience() != null) announcement.setTargetAudience(request.getTargetAudience());
        if (request.getTargetDepartmentIds() != null) announcement.setTargetDepartmentIds(request.getTargetDepartmentIds());
        if (request.getTargetEmployeeIds() != null) announcement.setTargetEmployeeIds(request.getTargetEmployeeIds());
        announcement.setExpiresAt(request.getExpiresAt());
        if (request.getIsPinned() != null) announcement.setIsPinned(request.getIsPinned());
        if (request.getSendEmail() != null) announcement.setSendEmail(request.getSendEmail());
        announcement.setAttachmentUrl(request.getAttachmentUrl());
        if (request.getRequiresAcceptance() != null) announcement.setRequiresAcceptance(request.getRequiresAcceptance());

        announcement = announcementRepository.save(announcement);
        log.info("Updated announcement: {}", announcement.getTitle());

        return AnnouncementDto.fromEntity(announcement);
    }

    @Transactional(readOnly = true)
    public Page<AnnouncementDto> getAllAnnouncements(Pageable pageable) {
        UUID tenantId = TenantContext.getCurrentTenant();
        UUID currentUserId = SecurityContext.getCurrentEmployeeId();
        return announcementRepository.findActiveAnnouncements(tenantId, LocalDateTime.now(), pageable)
                .map(a -> enrichAnnouncementDto(AnnouncementDto.fromEntity(a), currentUserId));
    }

    @Transactional(readOnly = true)
    public Page<AnnouncementDto> getActiveAnnouncements(UUID employeeId, Pageable pageable) {
        UUID tenantId = TenantContext.getCurrentTenant();

        // Get employee's department for filtering
        UUID employeeDepartmentId = null;
        boolean isManager = false;
        LocalDateTime joinDate = null;

        var employeeOpt = employeeRepository.findByIdAndTenantId(employeeId, tenantId);
        if (employeeOpt.isPresent()) {
            var employee = employeeOpt.get();
            employeeDepartmentId = employee.getDepartmentId();
            isManager = employee.getManagerId() == null ||
                    employeeRepository.findDirectReportsByManagerId(tenantId, employeeId).size() > 0;
            joinDate = employee.getJoiningDate() != null ? employee.getJoiningDate().atStartOfDay() : null;
        }

        List<AnnouncementRead> employeeReads = announcementReadRepository.findByEmployeeIdAndTenantId(employeeId, tenantId);

        Set<UUID> readAnnouncementIds = employeeReads.stream()
                .map(AnnouncementRead::getAnnouncementId)
                .collect(Collectors.toSet());

        Map<UUID, AnnouncementRead> readMap = employeeReads.stream()
                .collect(Collectors.toMap(AnnouncementRead::getAnnouncementId, r -> r));

        // Apply targeting filter
        final UUID finalDepartmentId = employeeDepartmentId;
        final boolean finalIsManager = isManager;
        final LocalDateTime finalJoinDate = joinDate;

        Page<Announcement> announcements = announcementRepository.findActiveAnnouncements(tenantId, LocalDateTime.now(), pageable);

        List<AnnouncementDto> filteredDtos = announcements.getContent().stream()
                .filter(a -> isAnnouncementVisibleToEmployee(a, employeeId, finalDepartmentId, finalIsManager, finalJoinDate))
                .map(a -> {
                    AnnouncementDto dto = AnnouncementDto.fromEntity(a);
                    dto.setIsRead(readAnnouncementIds.contains(a.getId()));
                    AnnouncementRead readRecord = readMap.get(a.getId());
                    if (readRecord != null) {
                        dto.setIsAccepted(readRecord.getIsAccepted());
                        dto.setAcceptedAt(readRecord.getAcceptedAt());
                    }
                    // Enrich with reaction status
                    return enrichAnnouncementDto(dto, employeeId);
                })
                .collect(Collectors.toList());

        return new PageImpl<>(filteredDtos, pageable, filteredDtos.size());
    }

    /**
     * Check if an announcement is visible to a specific employee based on targeting rules
     */
    private boolean isAnnouncementVisibleToEmployee(Announcement announcement, UUID employeeId,
                                                    UUID employeeDepartmentId, boolean isManager,
                                                    LocalDateTime joinDate) {
        if (announcement.getTargetAudience() == null) {
            return true; // Default to visible
        }

        switch (announcement.getTargetAudience()) {
            case ALL_EMPLOYEES:
                return true;

            case SPECIFIC_DEPARTMENTS:
                // Check if employee's department is in the target list
                if (employeeDepartmentId == null) {
                    return false;
                }
                Set<UUID> targetDepts = announcement.getTargetDepartmentIds();
                return targetDepts != null && targetDepts.contains(employeeDepartmentId);

            case SPECIFIC_EMPLOYEES:
                // Check if employee is in the target list
                Set<UUID> targetEmployees = announcement.getTargetEmployeeIds();
                return targetEmployees != null && targetEmployees.contains(employeeId);

            case MANAGERS_ONLY:
                return isManager;

            case NEW_JOINERS:
                // Consider employees who joined in the last 90 days as new joiners
                if (joinDate == null) {
                    return false;
                }
                LocalDateTime ninetyDaysAgo = LocalDateTime.now().minusDays(90);
                return joinDate.isAfter(ninetyDaysAgo);

            default:
                return true;
        }
    }

    @Transactional(readOnly = true)
    public List<AnnouncementDto> getPinnedAnnouncements() {
        UUID tenantId = TenantContext.getCurrentTenant();
        return announcementRepository.findPinnedAnnouncements(tenantId, LocalDateTime.now())
                .stream()
                .map(AnnouncementDto::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public AnnouncementDto getAnnouncementById(UUID announcementId, UUID employeeId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        AnnouncementDto dto = AnnouncementDto.fromEntity(announcement);

        if (employeeId != null) {
            announcementReadRepository.findByAnnouncementIdAndEmployeeIdAndTenantId(
                            announcementId, employeeId, tenantId)
                    .ifPresent(readRecord -> {
                        dto.setIsRead(true);
                        dto.setIsAccepted(readRecord.getIsAccepted());
                        dto.setAcceptedAt(readRecord.getAcceptedAt());
                    });
        }

        return dto;
    }

    @Transactional
    public void markAsRead(UUID announcementId, UUID employeeId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        // Check if already read
        if (announcementReadRepository.existsByAnnouncementIdAndEmployeeIdAndTenantId(
                announcementId, employeeId, tenantId)) {
            // Still record in generic view system (tracks repeat views)
            contentViewService.recordView(ContentType.ANNOUNCEMENT, announcementId, employeeId, "direct");
            return;
        }

        // Get announcement and increment read count
        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        announcement.incrementReadCount();
        announcementRepository.save(announcement);

        // Record the read
        AnnouncementRead read = AnnouncementRead.builder()
                .announcementId(announcementId)
                .employeeId(employeeId)
                .readAt(LocalDateTime.now())
                .build();
        read.setTenantId(tenantId);
        announcementReadRepository.save(read);

        // Also record in generic content view system
        contentViewService.recordView(ContentType.ANNOUNCEMENT, announcementId, employeeId, "direct");

        log.debug("Marked announcement {} as read by employee {}", announcementId, employeeId);
    }

    public void acceptAnnouncement(UUID announcementId, UUID employeeId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        // Get announcement
        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        // Check if requires acceptance
        if (!Boolean.TRUE.equals(announcement.getRequiresAcceptance())) {
            log.warn("Announcement {} does not require acceptance", announcementId);
            return;
        }

        // Find or create read record
        AnnouncementRead readRecord = announcementReadRepository
                .findByAnnouncementIdAndEmployeeIdAndTenantId(announcementId, employeeId, tenantId)
                .orElseGet(() -> {
                    AnnouncementRead newRead = AnnouncementRead.builder()
                            .announcementId(announcementId)
                            .employeeId(employeeId)
                            .readAt(LocalDateTime.now())
                            .build();
                    newRead.setTenantId(tenantId);
                    announcement.incrementReadCount();
                    return newRead;
                });

        // Check if already accepted
        if (Boolean.TRUE.equals(readRecord.getIsAccepted())) {
            return;
        }

        // Accept
        readRecord.accept();
        announcementReadRepository.save(readRecord);

        announcement.incrementAcceptedCount();
        announcementRepository.save(announcement);

        log.info("Announcement {} accepted by employee {}", announcementId, employeeId);
    }

    @Transactional
    public AnnouncementDto pinAnnouncement(UUID announcementId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        announcement.setIsPinned(true);
        announcement = announcementRepository.save(announcement);
        log.info("Pinned announcement: {}", announcementId);

        return AnnouncementDto.fromEntity(announcement);
    }

    @Transactional
    public AnnouncementDto unpinAnnouncement(UUID announcementId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        announcement.setIsPinned(false);
        announcement = announcementRepository.save(announcement);
        log.info("Unpinned announcement: {}", announcementId);

        return AnnouncementDto.fromEntity(announcement);
    }

    @Transactional
    public void deleteAnnouncement(UUID announcementId) {
        UUID tenantId = TenantContext.getCurrentTenant();

        Announcement announcement = announcementRepository.findByIdAndTenantId(announcementId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Announcement not found: " + announcementId));

        announcementRepository.delete(announcement);
        log.info("Deleted announcement: {}", announcementId);
    }

    /**
     * Enrich announcement DTO with user-specific data (hasReacted status)
     */
    private AnnouncementDto enrichAnnouncementDto(AnnouncementDto dto, UUID currentUserId) {
        if (dto.getWallPostId() != null && currentUserId != null) {
            boolean hasReacted = postReactionRepository.findByPostIdAndEmployeeId(
                    dto.getWallPostId(), currentUserId
            ).isPresent();
            dto.setHasReacted(hasReacted);
        }
        return dto;
    }
}
